docs/discussion/Cosign Upgrade 2024.md

```diff
@@ -0,0 +1,466 @@
+  - ~~Option 2: Store certs + keys on disk at well-known path~~
+    -  ~~`CertificateStore` will be refactored to store fetched certs from providers in a directory on disk~~
+      ```diff
+      - .ratify/certificate_store/
+      - <provider_name>-<namespace>/
+      -   keys/
+      -     key1.rsa
+      -   certificates/
+      -     cert1.pem
+      -     cert2.pem
+      ```
+    - ~~Configuration loading from JSON (CLI) or CRD reconcile loop will be responsible for deleting the directories of cert providers as they are removed~~
+    - ~~External plugins will have to have the `CertificateStore` config passed in as JSON input in order to create instance of the cert provider objects~~
+    - ~~External plugins will need to be namespace aware in order to know which directory to access~~
+    - ~~Pros:~~
+      - ~~Cosign plugin can manage accessing its own keys + certificates without executor having to coordinate~~
+      - ~~Other external plugins will not get unnecessary keys + certs provided in json input~~ 
+    - ~~Cons:~~
+      - ~~Keys + certs must be added/deleted off disk~~
+        - ~~What happens if keys + certs are left on disk due to Ratify failure?~~
+        - ~~Keys + certs will be able to be accessed by any process with filesystem access~~
+        - ~~CLI scenario will require fetching + deleting certs + keys off disk per invocation~~
+      - ~~Existing cert providers used by notation will need to be updated to read/write certs from disk~~
+      - ~~External plugins will need to be aware of namespace context in order to only access resources in namespaced directory~~
+  - ~~Option 3 (NOT recommended): Add certificates + keys to shared cache~~
+    - ~~Define two new cache keys partitioned by namespace~~
+    - ~~CANNOT use the ristretto in-memory cache as it is since external cosign process must be able to access it~~
+      - ~~This would require hard dependency on HA mode which is a bad experience. Why should a user need HA distributed cache just for Cosign?~~
+      - ~~Implement a new cache provider for cache sharing between processes on same host~~
+        - ~~NOT trivial~~
+        - ~~Require a new http cache server and clients~~
+    - ~~Cache provider initialization in cosign plugin implementation~~
+    - ~~One advantage is that not only will cert + key sharing be solved but also oras store credentials can easily shared which is currently not possible for external plugins~~
+- Extend Inline Certificate Provider
+  - add new `type` field to spec: "key" or "certificate". If not provided, default to "certificate" for backwards compatability. 
+- Extend Azure Key Vault Certificate Provider
+  - use `GetKeys` API to fetch keys from configured key vault
+  - add new `keys` field to spec
+- Multitenancy considerations
+  - Need a separate keys map for namespaced and global certificate store keys
+#### Renaming CertificateStore to KeyManagementSystem
```

Comment:
Will renaming CRD be a breaking change?

Reply:
We would have to introduce the new CRD KeyManagementSystem as a new and separate resource and then deprecate the existing one. I think removing full support for it would be breaking change... we should discuss this. The simplest option would be to keep the CertificateStore but the name is the only issue to consider.

Reply:
Is the new CRD name ` KeyManagementSystem` narrow to the Key Management System? If users use a managed signing service like AWS Signer or [Azure Code Signing service](https://techcommunity.microsoft.com/t5/security-compliance-and-identity/azure-code-signing-democratizing-trust-for-developers-and/ba-p/3604669), is this CRD applicable for non-KMS service?